docs/standard/native-interop/tutorial-custom-marshaller.md

```diff
@@ -0,0 +1,278 @@
+
+Marshalling of an `int` is unnecessary since representations are identical in unmanaged and managed. A `bool`'s binary representation isn't defined in .NET so we use its current value to define a zero and non-zero value in the unmanaged type. We then reuse our UTF-32 marshaller to convert our `string` field into a `uint*`.
+
+```csharp
+public static ErrorDataUnmanaged ConvertToUnmanaged(ErrorData managed)
+{
+    return new ErrorDataUnmanaged
+    {
+        Code = managed.Code,
+        IsFatal = (byte)(managed.IsFatalError ? 1 : 0),
+        Message = Utf32StringMarshaller.ConvertToUnmanaged(managed.Message),
+    };
+}
+```
+
+Recall we are defining this marshaller as an "in" so we must clean-up any allocations performed during the marshalling. The `int` and `bool` fields didn't allocate any memory, but the `Message` field did. We again reuse `Utf32StringMarshaller` to clean up the marshalled string.
+
+```csharp
+public static void Free(ErrorDataUnmanaged unmanaged)
+    => Utf32StringMarshaller.Free(unmanaged.Message);
+```
+
+Next, we are going to consider the case where multiple instances of `error_data` are returned. Typically returning multiple of a thing is done using a collection and those multiple things are called "elements". Since this will be returning multiple elements we call this the `MarshalMode.ElementOut` mode.
+
+```csharp
+namespace CustomMarshalling
+{
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedIn, typeof(ErrorDataMarshaller))]
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ElementOut, typeof(Out))]
+    internal static unsafe class ErrorDataMarshaller
+    {
+        //
+        // Other marshallers removed
+        //
+
+        public static class Out
+        {
+            public static ErrorData ConvertToManaged(ErrorDataUnmanaged unmanaged)
+                => throw new NotImplementedException();
+
+            public static void Free(ErrorDataUnmanaged unmanaged)
+                => throw new NotImplementedException();
+        }
+    }
+}
+```
+
+The conversion from `ErrorDataUnmanaged` to `ErrorData` is the inverse of what we did for the "in" mode above. Remember that we also need to "clean up" any allocations the unmanaged environment expected us to perform.
+
+For our managed to unmanaged "out" marshaller, we are going to do something special. The name of the data type we are marshalling is called `error_data` and .NET typically expresses errors as exceptions. However, some errors are more impactful than others. Errors identified as "fatal" usually indicate a catastrophic or unrecoverable error. Notice the `error_data` has a field to check if the error is fatal. Let's marshal an `error_data` into managed and if it is fatal, we will throw an exception rather than just converting it into an `ErrorData` and returning it.
```

Comment:
```suggestion
For the managed to unmanaged "out" marshaller, you're going to do something special. The name of the data type you're marshalling is called `error_data` and .NET typically expresses errors as exceptions. However, some errors are more impactful than others. Errors identified as "fatal" usually indicate a catastrophic or unrecoverable error. Notice the `error_data` has a field to check if the error is fatal. You'll marshal an `error_data` into managed code, and if it's fatal, you'll throw an exception rather than just converting it into an `ErrorData` and returning it.
```